### PR TITLE
explicit non-blocking RMA Operations

### PR DESCRIPTION
**Pull request by boehms**
This is in reference to #72

### CODE REVIEW

**Comment by boehms** [comment]
[main_spec.pdf](https://github.com/openshmem-org/specification/files/1226053/main_spec.pdf)

**Comment by abouteiller** [comment]
Typo on pp37 line 35: in the PUT routine description the text says 'get'

**Comment by abouteiller** [comment]
Jim and I expressed interest in adding a fence_nbe: 

```
putmem_nbe(req)
fence_nbe(req)
atomic_nbe(req)
wait(req)
```

would reduce the overhead when compared to a pure wait-based code snippet.

**Comment by abouteiller** [comment]
request_test/wait: 
rename to shmem(test|wait)_nbe
flip the flag to 1 when complete, 0 when incomplete (a la MPI_Test)
replicate the "note" about thread safety in test
expand the note about thread safety in a similar way to quiet (i.e. thread visibility and ordering must be user-enforced externally)

File: content/shmem_get_nbe.tex [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        be of type integer.}
+    \apiargument{IN/OUT}{req}{Either NULL, or a valid request object}
+\end{apiarguments}
+
+\apidescription{
+    The explicit non-blocking get routines provide a method for copying
+    a contiguous symmetric data object from a different \ac{PE} to a
+    contiguous data object on the local \ac{PE}. The routines return
+    after posting the operation.  The operation is considered complete
+    after a subsequent call to \FUNC{shmem\_wait\_request}. After the
```
Comment (manjugv) [inline content/shmem_get_nbe.tex]: Notes from WG discussion : \FUNC{shmem\_wait\_request} -> \FUNC{shmem\_request\_wait}

File: content/shmem_request_wait.tex [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+\apireturnvalues{
+    None.
+}
+
+\apinotes{ In a threaded environment \FUNC{shmem\_request\_wait} from
+a thread other than the thread, posting the operation is permitted.
+However, only one thread can call \FUNC{shmem\_request\_wait} for a
+particular request.
+}
```
Comment (manjugv) [inline content/shmem_request_wait.tex]: Notes from WG discussion: The paragraph should be reworked to reflect the ordering requirements between the operations and wait.

**Comment by boehms** [comment]
Adding most recent version

[main_spec.pdf](https://github.com/openshmem-org/specification/files/2006255/main_spec.pdf)

File: main_spec.tex [hunk 3]
```diff
@@ -138,6 +138,28 @@ \subsubsection{\textbf{SHMEM\_PUT\_NBI}}\label{subsec:shmem_put_nbi}
 \subsubsection{\textbf{SHMEM\_GET\_NBI}}\label{subsec:shmem_get_nbi}
 \input{content/shmem_get_nbi.tex}
 
+\color{ForestGreen}
+\subsection{Explicit Non-blocking Remote Memory Access Routines}\label{sec:rma_nbi}
```
Comment (BryantLam) [inline main_spec.tex]: For consistency, please rename the previous section from:
`\subsection{Non-blocking Remote Memory Access Routines}\label{sec:rma_nbi}`
to
`\subsection{Implicit Non-blocking Remote Memory Access Routines}\label{sec:rma_nbi}`

File: content/explicit_rma_intro.tex [hunk 4]
```diff
@@ -0,0 +1,12 @@
+The explicit nonblocking RMA routines provide interfaces for copying
+data from a contiguous local data object to a data object on a specified
+PE. The implicit and explicit nonblocking operations are similar in
+their semantics on return from the call .i.e., on return, they are not
```
Comment (BryantLam) [inline content/explicit_rma_intro.tex]: DocEdit: `call .i.e.,` --> `call; i.e.,`

File: content/explicit_rma_intro.tex [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+guaranteed to be completed either locally or remotely. However, they
+differ in their interfaces and how the operations are completed. The
+explicit nonblocking routines takes an extra argument called the request
```
Comment (BryantLam) [inline content/explicit_rma_intro.tex]: DocEdit: `takes` --> `take`

File: content/explicit_rma_intro.tex [hunk 6]
```diff
@@ -0,0 +1,12 @@
+The explicit nonblocking RMA routines provide interfaces for copying
+data from a contiguous local data object to a data object on a specified
+PE. The implicit and explicit nonblocking operations are similar in
+their semantics on return from the call .i.e., on return, they are not
+guaranteed to be completed either locally or remotely. However, they
+differ in their interfaces and how the operations are completed. The
```
Comment (BryantLam) [inline content/explicit_rma_intro.tex]: "However, they differ in their interfaces and how the operations are completed."
This sentence seems obvious. Please elaborate it or remove it.

File: content/explicit_rma_intro.tex (hunk 5 above)
Comment (BryantLam) [reply content/explicit_rma_intro.tex]: DocEdit: grammar for the rest of this paragraph below this point.

File: content/shmem_get_nbe.tex [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        be of type integer.}
+    \apiargument{IN/OUT}{req}{Either SHMEM\_REQUEST\_UNINITIALIZED, or a valid request object}
+\end{apiarguments}
+
+\apidescription{
+    The explicit non-blocking get routines provide a method for copying
+    a contiguous symmetric data object from a different \ac{PE} to a
+    contiguous data object on the local \ac{PE}. The routines return
+    after posting the operation.  The operation is considered complete
+    after a subsequent call to \FUNC{shmem\_wait\_nbe}. After the
```
Comment (BryantLam) [inline content/shmem_get_nbe.tex]: `subsequent call to \FUNC{shmem\_wait\_nbe}` ... with this same `shmem_request_t` handle object (some similar phrasing).

File: content/shmem_get_nbe.tex [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+\begin{apiarguments}
+    \apiargument{OUT}{dest}{Local data object to be updated.}
+    \apiargument{IN}{source}{Data object on the \ac{PE} identified by \VAR{pe}
+        that contains the data to be copied.  This data object must be remotely
+        accessible.}
+    \apiargument{IN}{nelems}{Number of elements in the \dest{} and \source{}
+        arrays. \VAR{nelems} must be of type \VAR{size\_t}.}
+    \apiargument{IN}{pe}{Identifier of the remote \ac{PE}. \VAR{pe} must
+        be of type integer.}
+    \apiargument{IN/OUT}{req}{Either SHMEM\_REQUEST\_UNINITIALIZED, or a valid request object}
```
Comment (BryantLam) [inline content/shmem_get_nbe.tex]: DocEdit: Remove "Either". Unnecessary.

`SHMEM_REQUEST_UNINITIALIZED` is only used here. It isn't mentioned in the API description. What's the effect of using this library constant?

If this is a new environment variable that needs to be in the library, please add to section Library Constants.

File: content/shmem_put_nbe.tex [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    data object must be remotely accessible.}
+    \apiargument{IN}{source}{Data object containing the data to be copied.}
+    \apiargument{IN}{nelems}{Number of elements in the \VAR{dest} and \VAR{source}
+    arrays. \VAR{nelems} must be of type \VAR{size\_t} for \Cstd. When using
+    \Fortran, it must be a constant, variable, or array element of default
+    integer type.}
+    \apiargument{IN}{pe}{\ac{PE} number of the remote \ac{PE}. \VAR{pe} must be
+    of type integer. When using \Fortran, it must be a constant, variable,
+    or array element of default integer type.}
+    \apiargument{IN/OUT}{req}{Either SHMEM\_REQUEST\_UNINITIALIZED, or a valid request object}
```
Comment (BryantLam) [inline content/shmem_put_nbe.tex]: Same comment as above. `SHMEM_REQUEST_UNINITIALIZED` is only used here. It isn't mentioned in the API description. Also, the API description mentions `NULL` but it isn't declared here as a valid input. (Or is `NULL` considered a valid request object? If so, that's confusing.)

**Comment by BryantLam** [comment]
Other comments from the Threads WG:
- Add example (preferably analogous to existing example in contexts section).

File: content/shmem_get_nbe.tex (hunk 7 above)
Comment (abouteiller) [reply content/shmem_get_nbe.tex]: and \FUNC{shmem\_test\_nbe}

File: content/shmem_put_nbe.tex [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    The explicit non-blocking put routines provide a method for copying a
+    contiguous symmetric data object from the local \ac{PE} to a contiguous data
+    object on a remote \ac{PE}.
+    The routines return after posting the operation. If the request
+    parameter is initialized to NULL, a new request will be created.
+    The delivery of data elements into the data object on the destination
+    \ac{PE} may occur in any order.
+    Furthermore, two successive put routines may deliver data out of
+    order. The operation is considered complete after a subsequent call
+    to \FUNC{shmem\_wait\_nbe} returns.
```
Comment (abouteiller) [inline content/shmem_put_nbe.tex]: and \FUNC{shmem_test_nbe} with a non-zero return flag

File: content/shmem_test_nbe.tex [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    zero.
+ }
+
+\apireturnvalues{
+    None.
+}
+
+\apinotes{
+In a threaded environment calling \FUNC{shmem\_test\_nbe} from
+a thread other than the thread, posting the operation is permitted.
```
Comment (abouteiller) [inline content/shmem_test_nbe.tex]: typo: extra coma

File: content/shmem_put_nbe.tex [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+\apidescription{
+    The explicit non-blocking put routines provide a method for copying a
+    contiguous symmetric data object from the local \ac{PE} to a contiguous data
+    object on a remote \ac{PE}.
+    The routines return after posting the operation. If the request
+    parameter is initialized to NULL, a new request will be created.
+    The delivery of data elements into the data object on the destination
+    \ac{PE} may occur in any order.
+    Furthermore, two successive put routines may deliver data out of
```
Comment (abouteiller) [inline content/shmem_put_nbe.tex]: seems to imply merged requests are still possible

File: content/shmem_wait_nbe.tex [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+\apireturnvalues{
+    None.
+}
+
+\apinotes{ In a threaded environment calling \FUNC{shmem\_wait\_nbe} from
+a thread other than the thread, posting the operation is permitted.
+However, only one thread must call \FUNC{shmem\_wait\_nbe} for a
+particular request. The \FUNC{shmem\_wait\_nbe} will return after all
+operation posted prior to the call are finished. It is the users
+responsibility that no new operations are posted to the request while it
```
Comment (abouteiller) [inline content/shmem_wait_nbe.tex]: seems to imply merged requests are still possible

File: content/shmem_wait_nbe.tex [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+\begin{apiarguments}
+    \apiargument{IN}{req}{Handle to the request object.}
+\end{apiarguments}
+
+\apidescription{
+    The \FUNC{shmem\_wait\_nbe} routine blocks until all outstanding operations on the request
+    are finished. Outstanding operations are operations posted before
+    wait was called.
+    There are no guarantees on the ordering of the outstanding
```
Comment (abouteiller) [inline content/shmem_wait_nbe.tex]: seems to imply merged requests are still possible

**Comment by boehms** [comment]
Adding most recent version

[main_spec.pdf](https://github.com/openshmem-org/specification/files/2138159/main_spec.pdf)

**Comment by BryantLam** [comment]
- Are merged handles still in scope? (I might have missed when this feature was dropped. The only mention on this ticket is in the commit message in a7c9b974f186ab3361b56cde1e0c78e1fb84a0be.)
    - What's the benefit of this API (without merged handles) vs. contexts? Needs some clarification of intent vs. contexts (when would a user choose to use this feature?).

- `shmem_handle_allocate` --> `shmem_handle_alloc`
    - Is alloc still needed without merged handles?

- Needs an example program demonstrating expected usage

- Constants need to be added to the Library Constants table

- `@\FuncParam@` for `<SIZE>` and `<TYPENAME>`

File: content/explicit_rma_intro.tex [hunk 15]
```diff
@@ -0,0 +1,11 @@
+The explicit non-blocking RMA routines provide interfaces for copying
+data from a contiguous local data object to a data object on a specified
+PE. The implicit and explicit non-blocking operations are similar in
+their semantics on return from the call; .i.e., on return, they are not
```
Comment (nspark) [inline content/explicit_rma_intro.tex]: There is an extra period preceding "i.e.". I suggest removing it entirely:
> ...their semantics on return from the call: on return, they are not...

File: content/explicit_rma_intro.tex [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+guaranteed to be completed either locally or remotely. However, they
+differ in their interfaces and how the operations are completed. The
+explicit non-blocking routines take an extra argument called the request
+handle that allows to track the progress of the RMA operation.
```
Comment (nspark) [inline content/explicit_rma_intro.tex]: > The explicit non-blocking routines take an extra argument called the request handle that allows **one** to track the progress of ~~the~~ **individual** RMA operation**s**.

File: content/shmem_put_nbe.tex [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    contiguous symmetric data object from the local \ac{PE} to a contiguous data
+    object on a remote \ac{PE}.
+    The routines return after posting the operation. If the request
+    parameter is initialized to SHMEM\_REQUEST\_UNINITIALIZED, a new
+    request will be created.
+    There are no guarantees to the order of operations if multiple successive operations are posted. 
+    The operation is considered complete after a subsequent call
+    to \FUNC{shmem\_wait\_nbe} for a given request handle returns, or a
+    call to \FUNC{shmem\_test\_nbe} for a given request handle returns
+    with a non-zero flag.
```
Comment (nspark) [inline content/shmem_put_nbe.tex]: For consistency with the `shmem_test_nbe` API, suggest:
> ...or a call to `shmem_test_nbe` for ~~a given~~ **the associated** request handle returns ~~with~~ a non-zero ~~flag~~ **status**.

File: content/shmem_test_nbe.tex [hunk 18]
```diff
@@ -0,0 +1,38 @@
+\apisummary{
+    The request test routine provides a method for querying the status
+    of the request.
+}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+void shmem_test_nbe(shmem_request_t *req, int *status);
```
Comment (nspark) [inline content/shmem_test_nbe.tex]: For consistency with `shmem_test`, I suggest that `shmem_test_nbe` return `status` directly from the function and not through a pointer argument.

File: content/shmem_test_nbe.tex [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+\end{Csynopsis}
+
+\begin{apiarguments}
+    \apiargument{IN}{req}{Handle to the request object.}
+    \apiargument{OUT}{status}{The status of the request.}
+\end{apiarguments}
+
+\apidescription{
+    The routine returns one via the status argument if the request is
+    finished, otherwise zero is returned.
```
Comment (nspark) [inline content/shmem_test_nbe.tex]: Suggest: "one" &#8594; 1 and "zero" &#8594; 0 for consistency with `shmem_test`

Comment (nspark) [reply content/shmem_test_nbe.tex]: Does "finished" imply local completion (i.e., the local buffer can be [re]used) or remote completion?

File: content/shmem_wait_nbe.tex [hunk 20]
```diff
@@ -0,0 +1,36 @@
+\apisummary{
+    The request wait routine provides a method for completing
+    outstanding operations on a request.
```
Comment (nspark) [inline content/shmem_wait_nbe.tex]: Does "completing" (and "completed", below) imply local completion (i.e., the local buffer can be [re]used) or remote completion?

File: content/shmem_put_nbe.tex [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+\apidescription{
+    The explicit non-blocking put routines provide a method for copying a
+    contiguous symmetric data object from the local \ac{PE} to a contiguous data
+    object on a remote \ac{PE}.
+    The routines return after posting the operation. If the request
+    parameter is initialized to SHMEM\_REQUEST\_UNINITIALIZED, a new
+    request will be created.
+    There are no guarantees to the order of operations if multiple successive operations are posted. 
+    The operation is considered complete after a subsequent call
+    to \FUNC{shmem\_wait\_nbe} for a given request handle returns, or a
```
Comment (nspark) [inline content/shmem_put_nbe.tex]: Remove the extra comma after "returns"

File: content/shmem_get_nbe.tex [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+\apidescription{
+    The explicit non-blocking get routines provide a method for copying
+    a contiguous symmetric data object from a different \ac{PE} to a
+    contiguous data object on the local \ac{PE}. The routines return
+    after posting the operation. If the request
+    parameter is initialized to SHMEM\_REQUEST\_UNINITIALIZED, a new
+    request will be created. There are no guarantees to the order of operations if multiple successive operations are posted.
+    The operation is considered complete after a subsequent call to \FUNC{shmem\_wait\_nbe} for a given
+    request handle returns, or after a call to \FUNC{shmem\_test\_nbe}
```
Comment (nspark) [inline content/shmem_get_nbe.tex]: Remove the extra comma after "returns"

File: content/shmem_test_nbe.tex [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+ }
+
+\apireturnvalues{
+    None.
+}
+
+\apinotes{
+In a threaded environment calling \FUNC{shmem\_test\_nbe} from
+a thread other than the thread posting the operation is permitted.
+The \FUNC{shmem\_wait\_nbe} will return the status of the operations
```
Comment (nspark) [inline content/shmem_test_nbe.tex]: I think you mean `shmem_test_nbe` here.

File: content/shmem_put_nbe.tex [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+\end{CsynopsisCol}
+
+\begin{apiarguments}
+    \apiargument{OUT}{dest}{Data object to be updated on the remote \ac{PE}. This
+    data object must be remotely accessible.}
+    \apiargument{IN}{source}{Data object containing the data to be copied.}
+    \apiargument{IN}{nelems}{Number of elements in the \VAR{dest} and \VAR{source}
+    arrays. \VAR{nelems} must be of type \VAR{size\_t} for \Cstd. When using
+    \Fortran, it must be a constant, variable, or array element of default
+    integer type.}
```
Comment (nspark) [inline content/shmem_put_nbe.tex]: Remove the Fortran reference

File: content/shmem_put_nbe.tex [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    \apiargument{IN}{pe}{\ac{PE} number of the remote \ac{PE}. \VAR{pe} must be
+    of type integer. When using \Fortran, it must be a constant, variable,
+    or array element of default integer type.}
```
Comment (nspark) [inline content/shmem_put_nbe.tex]: Remove the Fortran reference

**Comment by jdinan** [comment]
Closing due to inactivity, please reopen if interested.

**Pull request closed without merging**